.github/workflows/perform_build.yml

```diff
@@ -0,0 +1,75 @@
+# SPDX-License-Identifier: BSD-2-Clause
+# X-SPDX-Copyright-Text: (c) Copyright 2023 Advanced Micro Devices, Inc.
+
+name: "perform_build"
+
+on: push
+
+jobs:
+  extract_branch_info:
+    runs-on: ubuntu-latest
+    timeout-minutes: 1
+    outputs:
+      tcpdirectBranch: ${{ steps.tcpdirectBranch.outputs.branch }}
+      onloadBranch: ${{ steps.onloadBranch.outputs.branch }}
+    steps:
+      - name: tcpdirect
+        uses: actions/checkout@v4
+        with:
+          path: tcpdirect
+
+      - name: Extract TCPDirect branch name
+        id: tcpdirectBranch
+        run: echo "branch=${GITHUB_HEAD_REF:-${GITHUB_REF#refs/heads/}}" >> "$GITHUB_OUTPUT"
+
+      - name: Extract Onload branch name
+        id: onloadBranch
+        run: |
+          export tcpdirect_branch=${GITHUB_HEAD_REF:-${GITHUB_REF#refs/heads/}}
+          echo "branch=$(sed s/tcpdirect/onload/g <<< $tcpdirect_branch)" >> "$GITHUB_OUTPUT"
+
+  perform_build_and_test:
+    runs-on: ubuntu-latest
+    timeout-minutes: 20
+    needs: extract_branch_info
+    steps:
+      - name: tcpdirect
+        uses: actions/checkout@v4
+        with:
+          path: tcpdirect
+
+      - name: Install TCPDirect Deps
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y build-essential perl
+
+      - name: onload checkout
+        uses: actions/checkout@v4
+        with:
+          repository:  ${{ github.repository_owner }}/onload_internal
+          path: onload_internal
+          ssh-key: ${{ secrets.ONLOAD_INTERNAL_PRIVATE_KEY }}
+          ref: ${{needs.extract_branch_info.outputs.onloadBranch}}
+
+      - name: build
+        run: |
+          cd $GITHUB_WORKSPACE/tcpdirect
+          ONLOAD_TREE=$GITHUB_WORKSPACE/onload_internal NDEBUG=1 make -j $(nproc)
+
+      - name: shim
+        run: |
+          cd $GITHUB_WORKSPACE/tcpdirect
+          ONLOAD_TREE=$GITHUB_WORKSPACE/onload_internal NDEBUG=1 ZF_DEVEL=1 TEST_THREAD_NAME=zf make -j $(nproc) shim
+
+      - name: packetdrill checkout
+        uses: actions/checkout@v4
+        with:
+          repository:  ${{ github.repository_owner }}/packetdrill-tcpdirect
+          path: packetdrill-tcpdirect
+          ssh-key: ${{ secrets.PACKET_DRILL_PRIVATE_KEY }}
+          ref: tcpdirect-8.1
+
+      - name: tests
+        run: |
+          cd $GITHUB_WORKSPACE/tcpdirect
+          ONLOAD_TREE=$GITHUB_WORKSPACE/onload_internal NDEBUG=1 ZF_DEVEL=1 TEST_THREAD_NAME=zf make -j $(nproc) test
```

Comment:
I think I would probably like to have `-k` here, so we run through all tests even if one fails. Although I expect this might make two aspects a bit more tricky:
1. Would the run still "fail" if we ignore test failures and continue on to run everything?
2. How discoverable are the test failures - is there a summary at the end, or something that can be (relatively) easily searched for (e.g., "not ok")?

Reply:
This may be best as a future change for now, would rather this go in then sit here.

Reply:
Fair enough, definitely would like this in the future but agree that this can be iterated on.